crates/bevy_app/src/preferences.rs

```diff
@@ -0,0 +1,275 @@
+
+    use super::Preferences;
+
+    #[derive(Reflect, PartialEq, Debug)]
+    struct Foo(usize);
+
+    #[derive(Reflect, PartialEq, Debug)]
+    struct Bar(String);
+
+    fn get_registry() -> bevy_reflect::TypeRegistry {
+        let mut registry = bevy_reflect::TypeRegistry::default();
+        registry.register::<Foo>();
+        registry.register::<Bar>();
+        registry
+    }
+
+    #[test]
+    fn setters_and_getters() {
+        let mut preferences = Preferences::default();
+
+        // Set initial value
+        preferences.set(Foo(36));
+        assert_eq!(preferences.get::<Foo>().unwrap().0, 36);
+
+        // Overwrite with set
+        preferences.set(Foo(500));
+        assert_eq!(preferences.get::<Foo>().unwrap().0, 500);
+
+        // Overwrite with get_mut
+        *preferences.get_mut().unwrap() = Foo(12);
+        assert_eq!(preferences.get::<Foo>().unwrap().0, 12);
+
+        // Add new type of preference
+        assert!(preferences.get::<Bar>().is_none());
+        preferences.set(Bar("Bevy".into()));
+        assert_eq!(preferences.get::<Bar>().unwrap().0, "Bevy");
+
+        // Add trait object
+        preferences.set_dyn(Box::new(Bar("Boovy".into())));
+        assert_eq!(preferences.get::<Bar>().unwrap().0, "Boovy");
+
+        // Remove a preference
+        assert_eq!(*preferences.remove::<Foo>().unwrap(), Foo(12));
+    }
+
+    #[test]
+    fn init_exists() {
+        #[derive(Reflect, Clone, PartialEq, Debug)]
+        struct FooPrefs(String);
+
+        let mut app = App::new();
+        app.add_plugins(PreferencesPlugin);
+        app.update();
+        assert!(app.world().resource::<Preferences>().map.is_empty());
+    }
+
+    #[test]
+    fn startup_sets_value() {
+        #[derive(Reflect, Clone, PartialEq, Debug)]
+        struct FooPrefs(String);
+
+        let mut app = App::new();
+        app.add_plugins(PreferencesPlugin);
+        app.add_systems(Startup, |mut prefs: ResMut<Preferences>| {
+            prefs.set(FooPrefs("Initial".into()));
+        });
+        app.update();
+        assert_eq!(
+            app.world()
+                .resource::<Preferences>()
+                .get::<FooPrefs>()
+                .unwrap()
+                .0,
+            "Initial"
+        );
+    }
+
+    #[test]
+    fn serialization_round_trip() {
+        use bevy_reflect::serde::ReflectDeserializer;
+        use serde::{de::DeserializeSeed, Serialize};
+
+        let registry = get_registry();
+        let mut preferences = Preferences::default();
+        preferences.set(Foo(42));
+        preferences.set(Bar("Bevy".into()));
+
+        // Manually turn this into a valid JSON map. There is almost certainly a better way to
+        // express this if we want to make this part of the `Preferences` API as a blessed way to
+        // assemble a preferences file, but this is enough to get the file round tripping.
+        let mut output = String::new();
+        output.push('[');
```

Comment:
I'm a bit confused because the comment says "valid JSON map", but it's pushing a bracket, which starts a list.

I believe it's more idiomatic in JSON to use a map type for type/value pairs, where the type name is the key, rather than a list. This doesn't allow duplicate types, but that's the case here anyway.

Reply:
It's possible, by the way, to use serde to serialize the entire map. I have previously written a serde custom serializer that uses Bevy Reflect internally: https://github.com/viridia/panoply/blob/main/crates/panoply_exemplar/src/aspect_list.rs#L31

Reply:
Yeah, this is a list, the comment is wrong. This was a case of "it's 2AM and I want to build a serialization round tripping test". As I noted in that section, there is definitely a better way to do it, but this was enough to get it running and working. :)